### PR TITLE
[Merged by Bors] - bevy_reflect: Add `#[reflect(default)]` attribute for `FromReflect`

### PR DESCRIPTION
**Pull request by MrGVSV**
# Objective

Currently, `FromReflect` makes a couple assumptions:

* Ignored fields must implement `Default`
* Active fields must implement `FromReflect`
* The reflected must be fully populated for active fields (can't use an empty `DynamicStruct`)

However, one or both of these requirements might be unachievable, such as for external types. In these cases, it might be nice to tell `FromReflect` to use a custom default.

## Solution

Added the `#[reflect(default)]` derive helper attribute. This attribute can be applied to any field (ignored or not) and will allow a default value to be specified in place of the regular `from_reflect()` call. 

It takes two forms: `#[reflect(default)]` and `#[reflect(default = "some_func")]`. The former specifies that `Default::default()` should be used while the latter specifies that `some_func()` should be used. This is pretty much [how serde does it](https://serde.rs/field-attrs.html#default).

### Example

```rust
#[derive(Reflect, FromReflect)]
struct MyStruct {
  // Use `Default::default()`
  #[reflect(default)]
  foo: String,

  // Use `get_bar_default()`
  #[reflect(default = "get_bar_default")]
  #[reflect(ignore)]
  bar: usize,
}

fn get_bar_default() -> usize {
  123
}
```

### Active Fields

As an added benefit, this also allows active fields to be completely missing from their dynamic object. This is because the attribute tells `FromReflect` how to handle missing active fields (it still tries to use `from_reflect` first so the `FromReflect` trait is still required).

```rust
let dyn_struct = DynamicStruct::default();

// We can do this without actually including the active fields since they have `#[reflect(default)]`
let my_struct = <MyStruct as FromReflect>::from_reflect(&dyn_struct);
```

### Container Defaults

Also, with the addition of #3733, people will likely start adding `#[reflect(Default)]` to their types now. Just like with the fields, we can use this to mark the entire container as "defaultable". This grants us the ability to completely remove the field markers altogether if our type implements `Default` (and we're okay with fields using that instead of their own `Default` impls):

```rust
#[derive(Reflect, FromReflect)]
#[reflect(Default)]
struct MyStruct {
  foo: String,
  #[reflect(ignore)]
  bar: usize,
}

impl Default for MyStruct {
  fn default() -> Self {
    Self {
      foo: String::from("Hello"),
      bar: 123,
    }
  }
}

// Again, we can now construct this from nothing pretty much
let dyn_struct = DynamicStruct::default();
let my_struct = <MyStruct as FromReflect>::from_reflect(&dyn_struct);
```

Now if _any_ field is missing when using `FromReflect`, we simply fallback onto the container's `Default` implementation.

This behavior can be completely overridden on a per-field basis, of course, by simply defining those same field attributes like before.

### Related

* #3733
* #1395
* #2377

---

## Changelog

* Added `#[reflect(default)]` field attribute for `FromReflect`
  * Allows missing fields to be given a default value when using `FromReflect`
  * `#[reflect(default)]` - Use the field's `Default` implementation
  * `#[reflect(default = "some_fn")]` - Use a custom function to get the default value
* Allow `#[reflect(Default)]` to have a secondary usage as a container attribute
  * Allows missing fields to be given a default value based on the container's `Default` impl when using `FromReflect`

### CODE REVIEW

**Comment by MrGVSV** [comment]
With #3733 merged, I'm wondering: would also be good to make it so we use the struct's `Default` impl for unmarked ignored fields (if `#[reflect(Default)]` is defined)?

So for example, `value` would default to `123` instead of its usual `0` here:

```rust
#[reflect(Default)]
struct Foo {
  value: i32
}

impl Default for Foo {
  fn default() -> Self {
    Self {
      value: 123
    }
  }
}

let foo: Foo = FromReflect::from_reflect(dyn_foo_struct);
assert_eq!(123, foo.value);
```

This could be convenient for when defining custom `Default` impls. However, it could also be counter-intuitive and possibly footgunny since it would only be applicable for when `#[reflect(Default)]` is used.

Alternatively, we could define a sister attribute for the struct itself, like [serde does](https://serde.rs/container-attrs.html#default). This might result in more macro clutter, though...

**Review by PROMETHIA-27 — CHANGES_REQUESTED** [review]
Tests for correct behavior would probably also be a good idea. 

I like the changes overall, but mixing all of the cleanup changes with the actual feature additions does make it harder to review.

**Comment by MrGVSV** [comment]
> Tests for correct behavior would probably also be a good idea.

Lol was just adding some since I noticed that too.

> I like the changes overall, but mixing all of the cleanup changes with the actual feature additions does make it harder to review.

Would it be better if I split it up then? I like the changes because they make the end product a lot nicer to read/maintain. But if it's too difficult to review, then I can try to turn the cleanup changes into a followup PR.

**Comment by PROMETHIA-27** [comment]
I think it makes sense that if you reflect Default on a type, then creating an instance of that type from e.g. an empty dynamic struct should have all default values. So yeah, I think it's a good idea. I'm somewhat hesitant but I think it makes more sense than each value defaulting to its own default value. But also, it might cause a niche backwards incompatibility issue if that's done- I'm not sure if it's worth worrying about (or worth intentionally causing anyway) but good to keep in mind I think.
E.g. if I have the type
```
#[derive(Reflect)]
#[reflect(Default)]
struct Foo {
    x: i32
}
```
where x defaults to 12, this would suddenly change behavior from a 0'd x to 12 without changing the code. But reflect(Default) was also *just* added and hasn't been released in a version yet, so chances are no one is going to run into this.

**Comment by MrGVSV** [comment]
> where x defaults to 12, this would suddenly change behavior from a 0'd x to 12 without changing the code.

Yeah that was one of the concerns. Although, I think it makes sense that using `#[reflect(Default)]` would act much like serde's container `#[serde(default)]`. And thinking on it more, using an alternative mechanism (like our own container `#[reflect(default)]`) could result in weird/unexpected behavior.

So I think if a user lets the reflection system know it implements `Default`, then we should use it.

**Comment by MrGVSV** [comment]
Going to put this in draft form until #4712 is merged. We could just revert to the first commit of this PR and rebase from there, but I think it would just be easier to wait and see if the cleanup PR gets accepted/merged first so I can use some of the niceties added there.

If it's rejected or gets blocked for a long time, I'll go ahead and do the revert and rebase.

**Comment by MrGVSV** [comment]
Rebased and reopened!

**Comment by MrGVSV** [comment]
> Although, I think it makes sense that using `#[reflect(Default)]` would act much like serde's container `#[serde(default)]`

Added this behavior!

File: crates/bevy_reflect/bevy_reflect_derive/src/container_attributes.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/container_attributes.rs:22]:
```suggestion
// The traits listed below are not considered "special" (i.e. they use the `ReflectMyTrait` syntax)
```

File: crates/bevy_reflect/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/lib.rs:244]:
```suggestion
            // Use `get_bar_default()`
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Some small nits to clean up, then this LGTM. We'll need a second reviewer (@PROMETHIA-27?) before I can merge this though; it's not trivial and more eyes will help.

**Review by PROMETHIA-27 — APPROVED** [review]
LGTM!

**Pull request closed without merging**
